pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -120,13 +120,24 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
 		return framework.Queue, err
 	}
 	if isMatched {
-		logger.V(4).Info("node was created or updated, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
-		return framework.Queue, nil
+		if originalNode == nil || !pl.enableSchedulingQueueHint {
+			// The node was created or updated with QueueHints disabled (preCheck could have blocked some previous updates),
+			// so unblocking if only the new Pod is matching.
+			logger.V(4).Info("node was created or updated, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+			return framework.Queue, nil
+		}
+		// The operation is update and QueueHints are enabled (preCheck that was blocking some updates is removed),
+		// so we can narrow down the criteria to unmatch -> match changes (necessary affinity label was added to the node in this case).
+		wasMatched, err := requiredNodeAffinity.Match(originalNode)
+		if err != nil {
+			return framework.Queue, err
+		}
+		if pl.enableSchedulingQueueHint && isMatched && !wasMatched {
```

Comment:
```suggestion
		if isMatched && !wasMatched {
```
same here

Reply:
Even `if !wasMatched` is enough, as `isMatched` is already checked above.

Reply:
True